### PR TITLE
CB-11011 minimal AWS policies for Datalake, Datahub, Environment, Fre

### PR DESCRIPTION
**Pull request by topolyai5**
eipa and RDS. RDS is usable with dh or dl but others can be run as a 'standalone' policy for the role. These granulated policies support the roles to create and manage the dh, dl, env and freeipa instances on the AWS

See detailed description in the commit message.

### CODE REVIEW

File: cloud-aws/src/main/resources/definitions/env-minimal-policy.json [hunk 1]
```diff
@@ -0,0 +1,81 @@
+{
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/env-minimal-policy.json:1]: environment-minimal-policy

File: cloud-aws/src/main/resources/definitions/sdx-minimal-polcy.json [hunk 2]
```diff
@@ -0,0 +1,112 @@
+{
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/sdx-minimal-polcy.json:1]: datalake-minimal-policy

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/termination/StackPreTerminationAction.java [hunk 3]
```diff
@@ -59,7 +59,7 @@ protected void doExecute(StackTerminationContext context, TerminationEvent paylo
             StackPreTerminationFailed terminateStackResult = new StackPreTerminationFailed(payload.getResourceId(), new IllegalArgumentException(statusReason));
             sendEvent(context, StackTerminationEvent.PRE_TERMINATION_FAILED_EVENT.event(), terminateStackResult);
         } else {
-            putClusterToDeleteInProgressState(stack);
+//            putClusterToDeleteInProgressState(stack);
```
Comment (doktoric) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/termination/StackPreTerminationAction.java:62]: do we really want to do this ?

File: cloud-aws/src/main/resources/definitions/datahub-minimal-polcy.json [hunk 4]
```diff
@@ -0,0 +1,104 @@
+{
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-polcy.json:1]: Typo in file name: `polcy` -> `policy`.

File: cloud-aws/src/main/resources/definitions/datalake-minimal-polcy.json [hunk 5]
```diff
@@ -0,0 +1,114 @@
+{
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/datalake-minimal-polcy.json:1]: Typo in file name: `polcy` -> `policy`.

File: cloud-aws/src/main/resources/definitions/rds-minimal-polcy.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/rds-minimal-polcy.json:1]: Typo in file name: `polcy` -> `policy`.

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        "autoscaling:ResumeProcesses",
+        "autoscaling:CreateAutoScalingGroup",
+        "autoscaling:DeleteAutoScalingGroup"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "sts:AssumeRole"
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:65]: Why is this needed? This is what IDBroker is doing.

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      "Action": [
+        "iam:ListInstanceProfiles"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "sts:DecodeAuthorizationMessage",
+        "sts:AssumeRole"
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:69]: Why is this needed? This is what IDBroker is doing.

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json [hunk 9]
```diff
@@ -0,0 +1,82 @@
+{
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:1]: What is the decisive factor for what shall go into the freeipa policy and what into this one? I am slightly confused; we have EC2 here but not Autoscaling. Is this for creating network & DDB tables?

File: cloud-aws/src/main/resources/definitions/datahub-minimal-polcy.json [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        "cloudwatch:DescribeAlarms"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:PassRole",
+        "iam:ListInstanceProfiles"
+      ],
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-polcy.json:71]: Also needs the following one:
```
    {
      "Sid": "IdentityAccessManagementLimited",
      "Action": [
        "iam:CreateServiceLinkedRole"
      ],
      "Effect": "Allow",
      "Resource": [
        "arn:aws:iam::*:role/aws-service-role/*"
      ]
    },
```
Rationale: These special roles are needed for some services to act on behalf of the invoking user/role. For instance, `AWSServiceRoleForCloudWatchEvents` is needed to manipulate CW events (e.g. see https://jira.cloudera.com/browse/ENGESC-7128), and `AWSServiceRoleForAutoScaling` is needed to use EC2 Autoscaling, and perhaps `AWSServiceRoleForRDS` is needed for RDS usage. (Not sure about the latter, but it is also found in the `cdp-nonprod-datahub-policy.json` attachment of the ENGESC-7128 ticket.)

These roles can be preexisting in the target account, in which case we are happy as AWS will not have to recreate them, so there will be no complaints about `iam:CreateServiceLinkedRole`. On the other hand, trying to launch in a totally empty region or a brand new AWS account where these roles do not exist will manifest of lovely AWS errors. 😄

File: cloud-aws/src/main/resources/definitions/datalake-minimal-polcy.json [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:GetInstanceProfile",
+        "iam:SimulatePrincipalPolicy",
+        "iam:PassRole"
+      ],
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/datalake-minimal-polcy.json:72]: Also needs the `iam:CreateServiceLinkedRole` restricted to `arn:aws:iam::*:role/aws-service-role/*` resources. See my comment in the datahub policy.

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        "ec2:DeleteNatGateway",
+        "ec2:ImportKeyPair"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:ListInstanceProfiles"
+      ],
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:62]: Depending on what the purpose of this policy is, we might also need the `iam:CreateServiceLinkedRole` restricted to `arn:aws:iam::*:role/aws-service-role/*` resources. See my comment in the datahub policy.

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:PassRole"
+      ],
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:73]: Also needs the `iam:CreateServiceLinkedRole` restricted to `arn:aws:iam::*:role/aws-service-role/*` resources. See my comment in the datahub policy.

**Review by lajosrodek — COMMENTED** [review]
My unwanted 2 cents.

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json (hunk 9 above)
Comment (topolyai5) [reply cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:1]: This policy contains only vpc/subnet and dynamodb permissions.

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json (hunk 7 above)
Comment (topolyai5) [reply cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:65]: a leftover, I removed this

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json (hunk 8 above)
Comment (topolyai5) [reply cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:69]: I removed this

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+          "aws:RequestTag/Cloudera-Resource-Name": [
+            "crn:cdp:*"
+          ]
+        }
+      }
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "autoscaling:DeleteAutoScalingGroup",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:42]: delete operation should be restricted for cdp resources

File: cloud-aws/src/main/resources/definitions/rds-minimal-policy.json [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        "ec2:AuthorizeSecurityGroupEgress",
+        "ec2:DescribeAccountAttributes",
+        "ec2:DeleteSecurityGroup"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "rds:AddTagsToResource",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/rds-minimal-policy.json:34]: this should be cdp restricted right ?

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+          "aws:RequestTag/Cloudera-Resource-Name": [
+            "crn:cdp:*"
+          ]
+        }
+      }
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:DeleteTags",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:42]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+        "ec2:DescribeInstances",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:43]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+        "ec2:DescribeRegions",
+        "ec2:DescribeInstanceTypeOfferings",
+        "ec2:DescribeInstanceTypes",
+        "ec2:DescribeSubnets",
+        "ec2:DescribeKeyPairs",
+        "ec2:DescribeVpcs",
+        "ec2:DescribeImages",
+        "ec2:CreateSecurityGroup",
+        "ec2:DescribeSecurityGroups",
+        "ec2:DeleteSecurityGroup",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:54]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+        "ec2:CreateLaunchTemplate",
+        "ec2:DeleteLaunchTemplate",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:56]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+        "ec2:DescribeLaunchTemplates",
+        "ec2:DescribeAddresses",
+        "ec2:ReleaseAddress",
+        "ec2:RunInstances",
+        "ec2:CreateVolume",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:61]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+        "ec2:DescribeVolumes",
+        "ec2:AttachVolume",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:63]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+        "ec2:DescribeSecurityGroups",
+        "ec2:DeleteSecurityGroup",
+        "ec2:CreateLaunchTemplate",
+        "ec2:DeleteLaunchTemplate",
+        "ec2:DescribeLaunchTemplates",
+        "ec2:DescribeAddresses",
+        "ec2:ReleaseAddress",
+        "ec2:RunInstances",
+        "ec2:CreateVolume",
+        "ec2:DescribeVolumes",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:62]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+        "ec2:AttachVolume",
+        "ec2:DescribeRouteTables",
+        "ec2:TerminateInstances"
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:65]: cdp restricted

File: cloud-aws/src/main/resources/definitions/datalake-minimal-policy.json [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+          "aws:RequestTag/Cloudera-Resource-Name": [
+            "crn:cdp:*"
+          ]
+        }
+      }
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:DeleteTags",
```
Comment (doktoric) [inline cloud-aws/src/main/resources/definitions/datalake-minimal-policy.json:42]: cdp restricted

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json (hunk 9 above)
Comment (lajosrodek) [reply cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:1]: Super, got it.

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:PassRole"
+      ],
+      "Resource": [
+        "${ID_BROKER_ROLE}"
+      ]
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:96]: This is typically required for passing the EC2 instance profile to the EC2 service. For FreeIPA, this will always mean the Log Role; ID Broker (and its Assumer role) is not involved in any way.

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+        "cloudwatch:DeleteAlarms",
+        "cloudwatch:PutMetricAlarm",
+        "cloudwatch:DescribeAlarms"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:PassRole"
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:92]: We might also need the permission for `iam:GetRole` as noted in [this IAM doc page](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_passrole.html).

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json [hunk 27]
```diff
@@ -0,0 +1,106 @@
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:6]: Do not we also need `ec2:DeleteTags` (with the same `ec2:ResourceTag/Cloudera-Resource-Name` restriction)?

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+        "ec2:AllocateAddress",
+        "ec2:AssociateRouteTable",
+        "ec2:CreateRoute",
+        "ec2:DeleteRouteTable",
+        "ec2:DeleteVpcEndpoints",
+        "ec2:DisassociateRouteTable",
+        "ec2:ReleaseAddress",
+        "ec2:DeleteRoute",
+        "ec2:DeleteNatGateway",
+        "ec2:DeleteVpc",
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:89]: This comment only applies to the `Delete*` actions for new network resources: Do not we need them be restricted using `"ec2:ResourceTag/Cloudera-Resource-Name"`?

File: cloud-aws/src/main/resources/definitions/environment-minimal-policy.json [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+        "ec2:AuthorizeSecurityGroupIngress",
+        "sts:DecodeAuthorizationMessage",
+        "cloudformation:DescribeStacks",
+        "dynamodb:DeleteTable",
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/environment-minimal-policy.json:93]: Sanity check: This might look dangerous to customers, even though we will only execute it if the given DDB table does not yet exist at the time of environment launch. Do we apply any CDP specific resource tags to the DDB table? (Actually, I believe the DDB table is created by the S3A storage connector upon the first bucket access. Even so, we could potentially apply a custom tag to the table sometime after its creation.)

File: cloud-aws/src/main/resources/definitions/rds-minimal-policy.json [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:RevokeSecurityGroupEgress",
+        "ec2:AuthorizeSecurityGroupEgress",
+        "ec2:DescribeAccountAttributes",
+        "ec2:DeleteSecurityGroup"
+      ],
+      "Resource": "*"
+    },
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/rds-minimal-policy.json:60]: Why are these needed?

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:DeleteTags",
+        "ec2:AssociateAddress",
+        "ec2:StartInstances",
+        "ec2:StopInstances",
+        "ec2:AttachVolume",
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:11]: What about `ec2:DeleteVolume` and `ec2:DetachVolume`? Are these not needed?

File: cloud-aws/src/main/resources/definitions/datalake-minimal-policy.json [hunk 32]
```diff
@@ -0,0 +1,177 @@
...
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:DeleteTags",
+        "ec2:AssociateAddress",
+        "ec2:StartInstances",
+        "ec2:StopInstances",
+        "ec2:AttachVolume",
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/datalake-minimal-policy.json:11]: What about `ec2:DeleteVolume` and `ec2:DetachVolume`? Are these not needed?

File: cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+        "ec2:DeleteSecurityGroup",
+        "ec2:AssociateAddress",
+        "ec2:StartInstances",
+        "ec2:StopInstances",
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/freeipa-minimal-policy.json:10]: What about `ec2:TerminateInstances`? Is this also not needed?

File: cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+      "Action": [
+        "iam:PassRole"
+      ],
+      "Resource": "*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:ListInstanceProfiles",
+        "iam:SimulatePrincipalPolicy"
```
Comment (lajosrodek) [inline cloud-aws/src/main/resources/definitions/datahub-minimal-policy.json:116]: I wonder if this is really needed for Data Hub. Is this policy supposed to be a superset of the Data Lake policy? If not, `iam:SimulatePrincipalPolicy` is probably only used at environment and Data Lake level, but this shall be double checked in the code. 😃

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Some further comments.

File: cloud-aws/src/main/resources/definitions/rds-minimal-policy.json [hunk 35]
```diff
@@ -0,0 +1,83 @@
...
+            "crn:cdp:*"
+          ]
+        }
+      }
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "autoscaling:CreateAutoScalingGroup",
+        "autoscaling:DeleteAutoScalingGroup",
```
[the remark this answers is not in the document]
Comment (topolyai5) [reply cloud-aws/src/main/resources/definitions/rds-minimal-policy.json:25]: maybe yes, but we need the role to execute that, like the `CreateStack`, this role needs more role like vpc, subnet etc.

File: cloud-aws/src/main/resources/definitions/rds-minimal-policy.json [hunk 36]
```diff
@@ -0,0 +1,83 @@
...
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:PassRole"
+      ],
+      "Resource": [
+        "${ID_BROKER_ROLE}"
+      ]
+    }
```
[the remark this answers is not in the document]
Comment (topolyai5) [reply cloud-aws/src/main/resources/definitions/rds-minimal-policy.json:81]: it is similar because it creates an instance as well

File: cloud-aws/src/main/resources/definitions/rds-minimal-policy.json (hunk 30 above)
Comment (doktoric) [reply cloud-aws/src/main/resources/definitions/rds-minimal-policy.json:60]: new network scenario

Comment (lajosrodek) [reply cloud-aws/src/main/resources/definitions/rds-minimal-policy.json:60]: I see, thanks.

**Review by lajosrodek — APPROVED** [review]
Looks alright for a first starting point.
